File: src/proxy.rs
use axum::http::{HeaderMap, StatusCode};
use crate::config::Config;
use std::sync::Arc;

#[derive(Clone)]
pub struct AppState {
    pub config: Config,
    pub http_client: reqwest::Client,
    pub rag: Option<Arc<crate::rag::RagClient>>,
}

pub fn get_client_ip(headers: &HeaderMap, remote_addr: Option<std::net::SocketAddr>) -> String {
    if let Some(forwarded) = headers.get("X-Forwarded-For") {
        if let Ok(s) = forwarded.to_str() {
            return s.split(',').next().unwrap_or("unknown").trim().to_string();
        }
    }
    remote_addr.map(|a| a.ip().to_string()).unwrap_or_else(|| "unknown".to_string())
}

pub fn check_api_key(headers: &HeaderMap, expected_key: &str) -> Result<(), StatusCode> {
    if expected_key.is_empty() {
        return Ok(());
    }
    if let Some(auth) = headers.get(axum::http::header::AUTHORIZATION) {
        if let Ok(s) = auth.to_str() {
            if s == format!("Bearer {}", expected_key) {
                return Ok(());
            }
        }
    }
    if let Some(key) = headers.get("x-api-key") {
        if let Ok(s) = key.to_str() {
            if s == expected_key {
                return Ok(());
            }
        }
    }
    Err(StatusCode::UNAUTHORIZED)
}

File: proxy.go
package main

import (
	"net"
	"net/http"
	"strings"
)

// AppState holds the shared dependencies passed to every HTTP handler.
// Mirrors Rust's `proxy::AppState`.
type AppState struct {
	Config     Config
	HTTPClient *http.Client
	Rag        *RagClient // nil when RAG disabled (RagClient defined in rag.go)
}

// getClientIP extracts the originating client IP from a request.
// It prefers the first segment of the X-Forwarded-For header, falling back to
// the host portion of r.RemoteAddr, then "unknown".
func getClientIP(r *http.Request) string {
	if forwarded := r.Header.Get("X-Forwarded-For"); forwarded != "" {
		// split on ',', take the first segment, trim whitespace.
		first := forwarded
		if idx := strings.IndexByte(forwarded, ','); idx >= 0 {
			first = forwarded[:idx]
		}
		return strings.TrimSpace(first)
	}
	if r.RemoteAddr != "" {
		if host, _, err := net.SplitHostPort(r.RemoteAddr); err == nil {
			return host
		}
		return r.RemoteAddr
	}
	return "unknown"
}

// checkAPIKey reports whether the request is authorized.
// An empty expectedKey disables auth (always authorized). Otherwise the request
// must carry either "Authorization: Bearer <key>" or "x-api-key: <key>".
func checkAPIKey(r *http.Request, expectedKey string) bool {
	if expectedKey == "" {
		return true
	}
	if auth := r.Header.Get("Authorization"); auth != "" {
		if auth == "Bearer "+expectedKey {
			return true
		}
	}
	if key := r.Header.Get("x-api-key"); key != "" {
		if key == expectedKey {
			return true
		}
	}
	return false
}
